File: sdks/node/src/dashboard/policy.ts
// Auth policy for the dashboard API: which paths bypass the session gate,
// which methods need CSRF, and which actions are admin-only.

/** Exact paths that bypass the session/CSRF gate. */
export const PUBLIC_PATHS = new Set([
  "/api/auth/status",
  "/api/auth/login",
  "/api/auth/setup",
  "/api/auth/providers",
  "/health",
  "/readiness",
  "/metrics",
]);

/** Prefixes that bypass auth — OAuth paths carry a provider slot in the URL. */
export const PUBLIC_PATH_PREFIXES = ["/api/auth/oauth/start/", "/api/auth/oauth/callback/"];

/** State-changing paths a non-admin may call on their own session. */
const SELF_SERVICE_PATHS = new Set(["/api/auth/logout", "/api/auth/change-password"]);

/** Paths exempt from CSRF because no session exists yet. */
const CSRF_EXEMPT_PATHS = new Set(["/api/auth/login", "/api/auth/setup"]);

export function isPublicPath(path: string): boolean {
  return PUBLIC_PATHS.has(path) || PUBLIC_PATH_PREFIXES.some((p) => path.startsWith(p));
}

export function isStateChangingMethod(method: string): boolean {
  return method === "POST" || method === "PUT" || method === "DELETE" || method === "PATCH";
}

export function isCsrfExempt(path: string): boolean {
  return CSRF_EXEMPT_PATHS.has(path);
}

export function requiresAdmin(path: string, method: string): boolean {
  return isStateChangingMethod(method) && !SELF_SERVICE_PATHS.has(path);
}

File: sdks/java/src/main/java/org/byteveda/taskito/dashboard/auth/Policy.java
package org.byteveda.taskito.dashboard.auth;

import java.util.List;
import java.util.Set;
import org.byteveda.taskito.dashboard.support.DashboardError;

/**
 * The authorization gate: setup-required, authentication, CSRF, and RBAC.
 *
 * <p>RBAC is deliberately simple — every state-changing request is admin-only
 * except the self-service endpoints (logout, change-password); all reads are
 * allowed for any authenticated user. So admin vs viewer is purely write access.
 */
public final class Policy {
    private static final Set<String> PUBLIC_PATHS = Set.of(
            "/api/auth/status",
            "/api/auth/login",
            "/api/auth/setup",
            "/api/auth/providers",
            "/health",
            "/readiness",
            "/metrics");

    private static final List<String> PUBLIC_PREFIXES = List.of("/api/auth/oauth/start/", "/api/auth/oauth/callback/");

    private static final Set<String> SELF_SERVICE_PATHS = Set.of("/api/auth/logout", "/api/auth/change-password");

    private static final Set<String> CSRF_EXEMPT_PATHS = Set.of("/api/auth/login", "/api/auth/setup");

    private Policy() {}

    public static boolean isPublicPath(String path) {
        if (PUBLIC_PATHS.contains(path)) {
            return true;
        }
        for (String prefix : PUBLIC_PREFIXES) {
            if (path.startsWith(prefix)) {
                return true;
            }
        }
        return false;
    }

    public static boolean isStateChanging(String method) {
        return "POST".equals(method) || "PUT".equals(method) || "DELETE".equals(method) || "PATCH".equals(method);
    }

    private static boolean isCsrfExempt(String path) {
        return CSRF_EXEMPT_PATHS.contains(path);
    }

    private static boolean requiresAdmin(String path, String method) {
        return isStateChanging(method)
                && path.startsWith("/api/")
                && !isPublicPath(path)
                && !SELF_SERVICE_PATHS.contains(path);
    }

    /** Enforce the gate for an {@code /api/} request; throws on denial. */
    public static void authorize(String path, String method, RequestContext ctx, AuthStore store) {
        if (!isPublicPath(path) && store.countUsers() == 0) {
            throw DashboardError.serviceUnavailable("setup_required");
        }
        if (isPublicPath(path)) {
            return;
        }
        if (!ctx.authenticated()) {
            throw DashboardError.unauthorized("not_authenticated");
        }
        if (isStateChanging(method) && !isCsrfExempt(path) && !ctx.csrfValid()) {
            throw DashboardError.forbidden("csrf_failed");
        }
        if (requiresAdmin(path, method)
                && !AuthStore.ROLE_ADMIN.equals(ctx.session().role())) {
            throw DashboardError.forbidden("forbidden");
        }
    }
}
